File: test-models/kotlin-jackson/src/test/kotlin/test_service/models/ModelsTest.kt
package test_service.models

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.math.BigDecimal
import java.time.*
import java.util.*
import kotlin.test.*
import test_service.json.*

internal class ModelsTest {
    private val mapper = setupObjectMapper(jacksonObjectMapper())

    private fun <T> checkDeserialization(data: T, jsonStr: String, tClass: Class<T>) {
        val deserialized: T = mapper.readValue(jsonStr, tClass)
        assertEquals(data, deserialized)
    }

    private fun <T> checkSerialization(data: T, jsonStr: String) {
        val serialized = mapper.writeValueAsString(data)
        assertEquals(jsonStr, serialized)
    }

    private fun <T> check(data: T, jsonStr: String, tClass: Class<T>) {
        checkSerialization(data, jsonStr)
        checkDeserialization(data, jsonStr, tClass)
    }

    @Test
    fun objectModel() {
        val data = Message(42)
        val jsonStr = """{"field":42}"""
        check(data, jsonStr, Message::class.java)
    }

    @Test
    fun objectModelWrongFieldName() {
        assertFailsWith<Throwable> {
            mapper.readValue("""{"wrong_field":42}""", Message::class.java)
        }
    }

    @Test
    fun objectModelMissingField() {
        assertFailsWith<Throwable> {
            mapper.readValue("""{}""", Message::class.java)
        }
    }

    @Test
    fun nestedObject() {
        val data = Parent("the string", Message(123))
        val jsonStr = """{"field":"the string","nested":{"field":123}}"""
        check(data, jsonStr, Parent::class.java)
    }

    @Test
    fun objectFieldNotNull() {
        assertFailsWith<Throwable> {
            val jsonStr = """{"field":"the string","nested":null}"""
            mapper.readValue(jsonStr, Parent::class.java)
        }
    }

    @Test
    fun enumModel() {
        val data = EnumFields(Choice.SECOND_CHOICE)
        val jsonStr = """{"enum_field":"Two"}"""
        check(data, jsonStr, EnumFields::class.java)
    }

    @Test
    fun numericTypes() {
        val data = NumericFields(123, 1234, 1.23f, 1.23, BigDecimal("1.23"))
        val jsonStr = """{"int_field":123,"long_field":1234,"float_field":1.23,"double_field":1.23,"decimal_field":1.23}"""
        check(data, jsonStr, NumericFields::class.java)
    }

    @Test
    fun nonNumericTypes() {
        val data = NonNumericFields(
            true,
            "the string",
            UUID.fromString("123e4567-e89b-12d3-a456-426655440000"),
            LocalDate.parse("2019-11-30"),
            LocalDateTime.parse("2019-11-30T17:45:55")
        )
        val jsonStr =
            """{"boolean_field":true,"string_field":"the string","uuid_field":"123e4567-e89b-12d3-a456-426655440000","date_field":"2019-11-30","datetime_field":"2019-11-30T17:45:55"}"""
        check(data, jsonStr, NonNumericFields::class.java)
    }

    @Test
    fun arrayType() {
        val data = ArrayFields(listOf(1, 2, 3), listOf("one", "two", "three"))
        val jsonStr = """{"int_array_field":[1,2,3],"string_array_field":["one","two","three"]}"""
        check(data, jsonStr, ArrayFields::class.java)
    }

    @Test
    fun mapType() {
        val data = MapFields(
            hashMapOf("one" to 1, "two" to 2),
            hashMapOf("one" to "first", "two" to "second")
        )
        val jsonStr = """{"int_map_field":{"two":2,"one":1},"string_map_field":{"two":"second","one":"first"}}"""
        check(data, jsonStr, MapFields::class.java)
    }

    @Test
    fun optionalTypes() {
        val data = OptionalFields(123, "the string")
        val jsonStr = """{"int_option_field":123,"string_option_field":"the string"}"""
        check(data, jsonStr, OptionalFields::class.java)
    }

    @Test
    fun optionalTypesMissingFields() {
        val data = OptionalFields(null, null)
        val jsonStr = "{}"
        check(data, jsonStr, OptionalFields::class.java)
    }

    @Test
    fun jsonType() {
        val jsonField = """{"the_array":[1,"some string"],"the_object":{"the_bool":true,"the_string":"some value"},"the_scalar":123}"""
        val node: JsonNode = mapper.readTree(jsonField)
        val data = RawJsonField(node)
        val jsonStr = """{"json_field":{"the_array":[1,"some string"],"the_object":{"the_bool":true,"the_string":"some value"},"the_scalar":123}}"""
        check(data, jsonStr, RawJsonField::class.java)
    }

    @Test
    fun oneOfWrapper() {
        val canceled = OrderCanceled(UUID.fromString("123e4567-e89b-12d3-a456-426655440000"))
        val event: OrderEventWrapper = OrderEventWrapper.Canceled(canceled)
        val jsonStr = """{"canceled":{"id":"123e4567-e89b-12d3-a456-426655440000"}}"""
        check(event, jsonStr, OrderEventWrapper::class.java)
    }

    @Test
    fun oneOfWrapperItemNotNull() {
        assertFailsWith<Throwable> {
            val jsonStr = """{"canceled":null}"""
            mapper.readValue(jsonStr, OrderEventWrapper::class.java)
        }
    }

    @Test
    fun oneOfDiscriminatorTest() {
        val canceled = OrderCanceled(UUID.fromString("123e4567-e89b-12d3-a456-426655440000"))
        val event: OrderEventDiscriminator = OrderEventDiscriminator.Canceled(canceled)
        val jsonStr = """{"_type":"canceled","id":"123e4567-e89b-12d3-a456-426655440000"}"""
        check(event, jsonStr, OrderEventDiscriminator::class.java)
    }
}

File: tests/test-models/ts-superstruct/__tests__/models.test.ts
import * as t from '../test-service/superstruct'
import { checkEncodeDecode, datetime } from './util'

import {
  TMessage,
  Message,
  Choice,
  EnumFields,
  TEnumFields,
  Parent,
  TParent,
  NumericFields,
  TNumericFields,
  NonNumericFields,
  TNonNumericFields,
  ArrayFields,
  TArrayFields,
  MapFields,
  TMapFields,
  OptionalFields,
  TOptionalFields,
  OrderEventWrapper,
  TOrderEventWrapper,
  OrderEventDiscriminator,
  TOrderEventDiscriminator,
} from '../test-service/models';

describe('object', function() {
  let decoded: Message = {field: 123}
  let encoded = {'field': 123}
  checkEncodeDecode(TMessage, decoded, encoded)
  it('decode breaks', function() {
    expect(() => t.decode(TMessage, {})).toThrow();
  })
});

describe('enum fields', function() {
  let decoded: EnumFields = {enum_field: Choice.THIRD_CHOICE}
  let encoded = {'enum_field': 'Three'}
  checkEncodeDecode(TEnumFields, decoded, encoded)
  it('decode fail', function() {
    expect(() => t.decode(TEnumFields, {})).toThrowError()
  })
});

describe('nested types', function() {
  let decoded: Parent = {
    'field': 'the string',
    'nested': {'field': 123}
  }
  let encoded = {
    'field': 'the string',
    'nested': {'field': 123}
  }
  checkEncodeDecode(TParent, decoded, encoded)
  it('decode fail', function() {
    expect(() => t.decode(TParent, {})).toThrowError();
  })
});

describe('numeric fields', function() {
  let decoded: NumericFields = {
    int_field: 123,
    long_field: 123,
    float_field: 12.3,
    double_field: 12.3,
    decimal_field: 12.3,
  }
  let encoded = {
    int_field: 123,
    long_field: 123,
    float_field: 12.3,
    double_field: 12.3,
    decimal_field: 12.3,
  }
  checkEncodeDecode(TNumericFields, decoded, encoded)
  it('decode fail', function() {
    expect(() => t.decode(TNumericFields, {})).toThrowError();

  })
});

describe('non numeric fields', function() {
  let decoded: NonNumericFields = {
    boolean_field: true,
    string_field: 'some string',
    uuid_field: '123e4567-e89b-12d3-a456-426655440000',
    date_field: '2021-01-01',
    datetime_field: datetime(2021, 1, 2, 4, 54, 0),
  }
  let encoded = {
    boolean_field: true,
    string_field: 'some string',
    uuid_field: '123e4567-e89b-12d3-a456-426655440000',
    date_field: '2021-01-01',
    datetime_field: '2021-01-02T04:54:00',
  }
  checkEncodeDecode(TNonNumericFields, decoded, encoded)
});

describe('array fields', function() {
  let decoded: ArrayFields = {
    int_array_field: [1, 2, 3],
    string_array_field: ['one', 'two', 'three'],
  }
  let encoded = {
    int_array_field: [1, 2, 3],
    string_array_field: ['one', 'two', 'three'],
  }
  checkEncodeDecode(TArrayFields, decoded, encoded)
});

describe('map fields', function() {
  let decoded: MapFields = {
    int_map_field: {'one': 1, 'two': 2, 'three': 3},
    string_map_field: {'one': 'first', 'two': 'second', 'three': 'third'},
  }
  let encoded = {
    int_map_field: {'one': 1, 'two': 2, 'three': 3},
    string_map_field: {'one': 'first', 'two': 'second', 'three': 'third'},
  }
  checkEncodeDecode(TMapFields, decoded, encoded)
});

describe('optional fields', function() {
  let decoded: OptionalFields = {
    int_option_field: 123,
    string_option_field: 'the string',
  }
  let encoded = {
    int_option_field: 123,
    string_option_field: 'the string',
  }
  checkEncodeDecode(TOptionalFields, decoded, encoded)
});

describe('optional fields null', function() {
  let decoded: OptionalFields = {
    int_option_field: null,
    string_option_field: null,
  }
  let encoded = {
    int_option_field: null,
    string_option_field: null,
  }
  checkEncodeDecode(TOptionalFields, decoded, encoded)
});

describe('optional fields missing', function() {
  let decoded: OptionalFields = {
    int_option_field: undefined,
    string_option_field: undefined,
  }
  let encoded = {}
  checkEncodeDecode(TOptionalFields, decoded, encoded)
});

describe('oneof type - wrapped', function() {
  let decoded: OrderEventWrapper = { changed: { id: 'id123', quantity: 123 } }
  let encoded = { changed: {id: 'id123', quantity: 123} }
  checkEncodeDecode(TOrderEventWrapper, decoded, encoded)
  it('decode breaks', function() {
    expect(() => t.decode(TOrderEventWrapper, { created: {} })).toThrowError();
  })
});

describe('oneof type - discriminated', function() {
  let decoded: OrderEventDiscriminator = { _type: 'changed', id: 'id123', quantity: 123 }
  let encoded = { _type: 'changed', id: 'id123', quantity: 123 }
  checkEncodeDecode(TOrderEventDiscriminator, decoded, encoded)
  it('decode breaks', function() {
    expect(() => t.decode(TOrderEventDiscriminator, { _type: 'changed' })).toThrowError('Expected the value to satisfy a union');
  })
});